File: Programmers/SeungDo/3주차(Stack, Queue)/올바른 괄호.cpp
#include<string>
#include<stack>
using namespace std;
// 문자열s의 길이만큼 for문을 돌려 스택이 비었거나 괄호완성이 안되면 push 괄호완성이 되면 top값 pop하는 방식으로 풀이해봤습니다.
bool solution(string s)
{
    stack <char> prime; //char형 스택 선언
    for (int i = 0; i < s.size(); i++) {
        if (prime.empty() == 1 || prime.top() != '(' || s[i] != ')') prime.push(s[i]); // 스택이 비었거나 탑 값이 '('가 아니거나 s[i]값이 ')'가 아니면 s[i]값 푸쉬
        else prime.pop(); // 스택의 탑값이 '(' 이고 푸쉬할 값이 ')'일때 top값 팝
    }
    if(prime.empty() == 1) return true; // 스택이 비었다면 true리턴
    if(prime.empty() == 0) return false; // 스택에 값이 남아있다면 false리턴
}

File: Programmers/JiHo/3주차(Stack, Queue)/올바른 괄호.py
'''
시간복잡도 O(N)
고려해야하는 요소가 '('또는')' 두개뿐이므로 직접 요소를 저장할 필요가 없다고 생각 -> 정수형으로 저장
( = -1, ) = +1으로 저장할 경우 ()와 같이 정확하게 닫혔을때 0
())와 같이 괄호가 닫히지 않았을 때 0보다 커지게 됨
'''
def solution(s):
    stack = 0 #상태저장용 스택
    for i in s :
        if i == "(" : stack -= 1 # (이면 -1
        elif i == ")" : stack += 1 # )이면 +1
        if stack > 0 : return False #매번 요소마다 stack이 0을 넘는지 확인
    if stack != 0 : return False #모든요소 검사시 열려있는 괄호가 있는지 확인
    return True #위에서 안걸렸을경우 정상괄호
